### DISCUSSION TITLE
What does your Android Development Setup look like?

### DISCUSSION THREAD

**Discussion by phusting** [apache/cordova#358] [General]
I'm rebuilding my development machine completely because it seems my Android studio must be corrupted or something.   I would love to know what your development setup is to build with CLI11 and Android11.  I'm using Windows 10 too.

Gradle version: 
Java version and which one, who you got it from: 
Android Studio Verson or Build Date :

in Android Studio,
SDK Platforms
Android SDK Build-Tools

any special settings in your config.xml for building

Thanks all 🤓

**Comment by breautek** [comment]
I'm on Ubuntu 22, so my environment is going to be different but

```
gradle --version

------------------------------------------------------------
Gradle 7.5.1
------------------------------------------------------------

Build time:   2022-08-05 21:17:56 UTC
Revision:     d1daa0cbf1a0103000b71484e1dbfe096e095918

Kotlin:       1.6.21
Groovy:       3.0.10
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          11.0.17 (Ubuntu 11.0.17+8-post-Ubuntu-1ubuntu222.04)
OS:           Linux 5.15.0-56-generic amd64
```

```
java --version
openjdk 11.0.17 2022-10-18
OpenJDK Runtime Environment (build 11.0.17+8-post-Ubuntu-1ubuntu222.04)
OpenJDK 64-Bit Server VM (build 11.0.17+8-post-Ubuntu-1ubuntu222.04, mixed mode, sharing)

javac --version
javac 11.0.17
```

```
Android Studio Dolphin | 2021.3.1
Build #AI-213.7172.25.2113.9014738, built on August 31, 2022
Runtime version: 11.0.13+0-b1751.21-8125866 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Linux 5.15.0-56-generic
GC: G1 Young Generation, G1 Old Generation
Memory: 4096M
Cores: 16
Registry:
    external.system.auto.import.disabled=true
    ide.text.editor.with.preview.show.floating.toolbar=false
    ide.instant.shutdown=false
    ide.images.show.chessboard=true


Current Desktop: ubuntu:GNOME
```

Build tools depends on the cordova android version... Current version requires Build Tools 32 and I'm using 32.0.0.
SDK Platform is SDK 32 (Whichever compile SDK you're using, which by default is 32).

Additionally, my environment variables looks like:

```
export JAVA_HOME="/usr/lib/jvm/java-11-openjdk-amd64"
export ANDROID_SDK_ROOT="/development/android/"
export ANDROID_HOME=$ANDROID_SDK_ROOT

export PATH="$PATH:$ANDROID_SDK_ROOT/cmdline-tools/7.0/bin"
export PATH="$PATH:$ANDROID_SDK_ROOT/platform-tools/"
export PATH="$PATH:$ANDROID_SDK_ROOT/emulator/"

export BUILD_TOOLS_VERSION="32.0.0"
export PATH="$PATH:$ANDROID_SDK_ROOT/build-tools/$BUILD_TOOLS_VERSION/"

export PATH="/usr/lib/jvm/java-11-openjdk-amd64/bin:$PATH"
```

**Reply by phusting** [reply]
Thank you for sharing, this is helpful.  I swear Android Studio did  the update and then everything broke...  I'm going to try an older version of Android Studio.

Again, thank you

**Reply by breautek** [reply]
If you open the android project in Android Studio, it is possible it will prompt/bug you to update things and if you do it may break the Cordova project as updating dependencies is usually only half of the story.

Cordova will generate the android project using specific dependency versions it supports. So if you open the project in Android Studio, avoid updating things like the Android Gradle Plugin (AGP).

Additionally, often times Android Studio needs a specific AGP installed for it to manage your android project, so having a newer/latest Android Studio may not work with the android project that Cordova generates.

**Comment by ndrake** [comment]
I'm developing on a Mac and my setup is as follows:

* Gradle 7.6
* Java 11: from OpenJDK (`brew install openjdk@11`)
* Android Studio: Chipmunk | 2021.2.1 Patch 2
* SDK Platforms: versions 29 up to and including 33
* SDK Build Tools: 32.0.0 and 30.0.3
* We have `<preference name="android-targetSdkVersion"   value="32" />` in our config.xml

Cordova Info:
```
Cordova Packages:

    cli: 11.0.0
        common: 4.0.2
        create: 4.0.0
        lib: 11.0.0
            common: 4.0.2
            fetch: 3.0.1
            serve: 4.0.0
Environment:

    OS: macOS Ventura 13.0.1 (22A400) (darwin 22.1.0) arm64
    Node: v16.15.0
    npm: 8.5.5

Cordova Android 11.0.0
Cordova iOS 6.2.0
```

**Reply by phusting** [reply]
You know, I've been developing on Windows for years.  But I'm going to fire up my MAC and try building for Android over there.  I do my Development, Browser and Android builds on windows and the copy everything over to MAC.  Maybe, the MAC should be my new machine and everything done there?

**Reply by phusting** [reply]
What IDE do you use on the Mac?

**Reply by breautek** [reply]
UNIX machines are the way to go for development imo.

I've used them all, Windows, Linux, and Macs. Linux is my choice for general development, but I use Macs when I need to (e.g. working with iOS stuff).

I would definitely recommend using a Mac for development if possible over Windows, unless if you have a specific need to develop on Windows.

**Reply by ndrake** [reply]
I use [VSCode](https://code.visualstudio.com/) for editing.

**Reply by phusting** [reply]
I was born on Windows back in the late 80's.  Tough to move away from it.  :shushing_face:

**Reply by phusting** [reply]
> I'm developing on a Mac and my setup is as follows:
> 
> * Gradle 7.6
> * Java 11: from OpenJDK (`brew install openjdk@11`)
> * Android Studio: Chipmunk | 2021.2.1 Patch 2
> * SDK Platforms: versions 29 up to and including 33
> * SDK Build Tools: 32.0.0 and 30.0.3
> * We have `<preference name="android-targetSdkVersion"   value="32" />` in our config.xml
> 
> Cordova Info:
> 
> ```
> Cordova Packages:
> 
>     cli: 11.0.0
>         common: 4.0.2
>         create: 4.0.0
>         lib: 11.0.0
>             common: 4.0.2
>             fetch: 3.0.1
>             serve: 4.0.0
> Environment:
> 
>     OS: macOS Ventura 13.0.1 (22A400) (darwin 22.1.0) arm64
>     Node: v16.15.0
>     npm: 8.5.5
> 
> Cordova Android 11.0.0
> Cordova iOS 6.2.0
> ```

Does all this work on a M1 Mac?

**Reply by ndrake** [reply]
Yep, I'm using an M1.  I've built/published multiple apps using the M1 with the latest Cordova versions.

**Reply by phusting** [reply]
Thank you.  I'm thinking about buying a Mini for just app development.  headless box I could remote desktop into.

**Reply by breautek** [reply]
Mini is also on my wishlist. Just a small power box for app development, but I have a KVM switch so that I can apply my usb devices and monitor to it when I want to use it.

**Comment by phusting** [comment]
I rebuild my Windows computer and using @ndrake above and now getting this new error.  It's gradle related, but no idea what this means

**FAILURE: Build failed with an exception.**
* What went wrong:
Execution failed for task ':app:processReleaseResources'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade
   > Android resource linking failed
     C:\Users\Patrick\.gradle\caches\transforms-2\files-2.1\42f4abcd1335f10590e79520ce40d264\core-1.7.0\res\values\values.xml:105:5-114:25: AAPT: error: resource android:attr/lStar not found.

**Reply by breautek** [reply]
AAPT issues are usually related to using the wrong build tools version.

**Comment by phusting** [comment]
For fun, removed _cordova platform android_ and did a reinstall _cordova platform add android@10_ and noticed that a "Downloading https://services.gradle.org/distributions/gradle-7.1.1-all.zip"   Where is this getting installed?   I have gradle installed on my computer which is 7.6 and reflected in the PATH, but I'm wondering if they are getting confused with this getting installed somewhere??

**Reply by phusting** [reply]
![image](https://user-images.githubusercontent.com/12736097/206578762-2b572add-f46c-4969-9189-e0b03b001a27.png)

**Reply by breautek** [reply]
The system Gradle is used to download a gradle wrapper.

A gradle wrapper is a specific version of gradle a particular project uses. This is so you can have many different projects on your computer, each potentially configured to a different version of Gradle without it conflicting with another project.

However cordova-android@11 should be using version `7.4.2` by default.

Gradle `7.1.1` was used by cordova-android@10.

**Comment by phusting** [comment]
I really want to get my Windows machine working again.   Now this is the latest message I'm going when running the build command.  When I pull in into Android studio and run it there, same message.   If I do a new/simple helloworld app, it builds fine.  Do you think it could be a plugin or some weird tag in my config.xml doing this?

Manifest merger failed : Apps targeting Android 12 and higher are required to specify an explicit value for `android:exported` when the corresponding component has an intent filter defined. See https://developer.android.com/guide/topics/manifest/activity-element#exported for details.

CORDOVA BUILD ANDROID
Warning: unexpected element (uri:"", local:"base-extension"). Expected elements are <{}codename>,<{}layoutlib>,<{}api-level>

> Task :app:processDebugMainManifest FAILED
C:\Users\Patrick\Documents\Projects\TheEquestrian\EquestrianApp\Equestrian App\platforms\android\app\src\main\AndroidManifest.xml Error:
        Apps targeting Android 12 and higher are required to specify an explicit value for `android:exported` when the corresponding component has an intent filter defined. See https://developer.android.com/guide/topics/manifest/activity-element#exported for details.

See http://g.co/androidstudio/manifest-merger for more information about the manifest merger.

**Reply by breautek** [reply]
cordova-android has `android:exported` explicitly set as of version [10.1.2](https://github.com/apache/cordova-android/blob/master/RELEASENOTES.md#1012-apr-11-2022)

But it's possible that plugins introduce other activities that also requires the `android:exported`, in which case the plugin maintainers needs to update their plugins.

**Reply by phusting** [reply]
So would a good test is remove all the plugins, re-add one at a time, build and see what happens?  Then add another and build?

**Reply by breautek** [reply]
> So would a good test is remove all the plugins, re-add one at a time, build and see what happens? Then add another and build?

That would be a bruteforce way... effective if you aren't sure what you're looking for. I would remove all the plugins and remove/re-add the platform as well. At this point a build should pass (even if you can't run your app), then you can start adding plugins one by one until the build breaks.

The reasoning for remove/re-add the platform after removing all the plugins is to ensure that you're starting from a clean slate... it's possible a plugin doesn't clean up after itself on uninstall and if that were the case, you may have potentially a broken build even if you uninstall the offending plugin.

**Reply by phusting** [reply]
Okay, I'm working through this right now.  So far half the plugins are building, the other have to install and test still

  Question is cordova-plugin-androidx-adapter or cordova-plugin-compat still needed in cordova-android 11?

**Reply by breautek** [reply]
I don't know what `cordova-plugin-compat` is but `cordova-plugin-androidx-adapter` would be required if you use any plugins that still uses the old android support libraries.

`cordova-plugin-androidx-adapter` adapts the old support libraries and transforms them to the androidx support library equivalent.

**Reply by phusting** [reply]
I found the issues!!!!!   It was these two plugins causing my builds to fail.  Anyone got a recommendation for a new QR code reader plugin?  The image cropper is not big deal losing that.

cordova-plugin-k-imagecropper 1.2.0 "KImageCropper"
cordova-plugin-qr-barcode-scanner 8.0.3 "BarcodeScanner"

**Comment by gwhenne** [comment]
A way to side step issues with build stacks is to use a build service. There are several which support Cordova. It then becomes the problem of the maintainers of the services to keep everything up to date and working together.

There are constant changes to hardware, operating systems, versions of build tools, Cordova, plugins, stores, etc. These happen on both your own dev system and on devices. It can be a lot of work keeping up with. If you're like many devs, you only build executables when you are ready to release. If it's been a few months since you used your build stack, it almost certainly will have broken in the meantime.

A happy side effect is that you don't need a Mac to build iOS apps. Upload your project to the service and it will return the appropriate IPA.

Disclosure: I'm one of the maintainers of [VoltBuilder](https://volt.build/), which is such a service. I developed Cordova apps both locally and with PhoneGap since they started, so I know the pain.
